prboom2/src/p_maputl.c

```diff
@@ -725,6 +725,195 @@ dboolean P_PathTraverse(fixed_t x1, fixed_t y1, fixed_t x2, fixed_t y2,
   return P_TraverseIntercepts(trav, FRACUNIT);
 }
 
+//
+// RoughBlockCheck
+// [XA] adapted from Hexen -- used by P_RoughMonsterSearch
+//
+
+static mobj_t *RoughBlockCheck(mobj_t *mo, int index, angle_t fov)
+{
+  mobj_t *link;
+  mobj_t *master;
+  angle_t angle, minang, maxang;
+
+  // pre-calculate fov check stuff since it
+  // won't change during the blocklinks loop
+  if (fov > 0) {
+    minang = mo->angle - fov / 2;
+    maxang = mo->angle + fov / 2;
+  }
+
+  link = blocklinks[index];
+  while (link)
+  {
+    // skip non-shootable actors
+    if (!(link->flags & MF_SHOOTABLE))
+    {
+      link = link->bnext;
+      continue;
+    }
+
+    // skip other players in co-op
+    if (netgame && !deathmatch && link->player)
+    {
+      link = link->bnext;
+      continue;
+    }
+
+    // skip the projectile's owner
+    if (link == mo->target)
+    {
+      link = link->bnext;
+      continue;
+    }
+
+    // skip friendly actors if owner is friendly (or a player)
+    if ((link->flags & MF_FRIEND) && mo->target != NULL &&
+      ((mo->target->flags & MF_FRIEND) || mo->target->player))
+    {
+      link = link->bnext;
+      continue;
+    }
+
+    // skip actors outside of specified FOV
+    // [XA] code borrowed from EE; thanks Quas :D
+    if (fov > 0)
+    {
+      angle = R_PointToAngle2(mo->x, mo->y, link->x, link->y);
+      // if the angles are backward, compare differently
+      if((minang > maxang) ? angle < minang && angle > maxang
+                           : angle < minang || angle > maxang)
+      {
+        link = link->bnext;
+        continue;
+      }
+    }
+
+    // skip actors not in line of sight
+    if (!P_CheckSight(mo, link))
+    {
+      link = link->bnext;
+      continue;
+    }
+
+    // all good! return it.
+    return link;
+  }
+
+  // couldn't find a valid target
+  return NULL;
+}
+
+//
+// P_RoughMonsterSearch
+// Searches though the surrounding mapblocks for monsters/players
+//
+// distance is in MAPBLOCKUNITS
+
+mobj_t *P_RoughMonsterSearch(mobj_t *mo, angle_t fov, int distance)
```

Comment:
Since this isn't searching for just monsters, let's call it `P_RoughTargetSearch` for consistency with `P_LookForTargets`.

Reply:
Works for me :)